src/v1/parser.rs

```diff
@@ -0,0 +1,814 @@
+/// v1 format reader
+pub struct Parser<R: BufRead> {
+    header: Header,
+    reader: R,
+}
+
+impl<R: BufRead> Parser<R> {
+    /// Creates a directory signature parser (format v1)
+    /// Tries to parse header
+    pub fn new(mut reader: R) -> Result<Parser<R>, ParseError> {
+        let mut header_line = vec!();
+        read_line(&mut reader, &mut header_line)?;
+        let header = Header::parse(&header_line).context(1)?;
+        Ok(Parser {
+            header: header,
+            reader: reader,
+        })
+    }
+
+    /// Returns parsed `Header`
+    pub fn get_header(&self) -> Header {
+        self.header.clone()
+    }
+
+    /// Creates iterator over directory signature entries
+    pub fn iter(&mut self) -> EntryIterator<R> {
+        EntryIterator::new(&mut self.reader,
+            self.header.hash_type, self.header.block_size)
+    }
+}
+
+pub struct EntryIterator<'a, R: 'a + BufRead> {
+    reader: &'a mut R,
+    hash_type: HashType,
+    block_size: u64,
+    current_dir: PathBuf,
+    current_row_num: usize,
+}
+
+impl<'a, R: BufRead> EntryIterator<'a, R> {
+    fn new(reader: &'a mut R, hash_type: HashType, block_size: u64)
+        -> EntryIterator<R>
+    {
+        EntryIterator {
+            reader: reader.by_ref(),
+            hash_type: hash_type,
+            block_size: block_size,
+            current_dir: PathBuf::new(),
+            current_row_num: 1,
+        }
+    }
+}
+
+impl<'a, R: BufRead> Iterator for EntryIterator<'a, R> {
+    type Item = Result<Entry, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut buf = vec!();
+        let is_last = itry!(read_line(self.reader.by_ref(), &mut buf));
+        let row = &buf[..];
+        self.current_row_num += 1;
+        let entry = itry!(Entry::parse(
+                row, &self.current_dir, self.hash_type, self.block_size)
+            .context(self.current_row_num));
+        match entry {
+            None => {
+                let _footer = itry!(Footer::parse(row, self.hash_type)
+                    .context(self.current_row_num));
+                if is_last {
+                    return Some(Err(ParseError::Parse(
+                        format!("Footer must be ended by a newline"),
+                        self.current_row_num)));
+                }
+                let mut test_buf = [0; 1];
+                if itry!(self.reader.read(&mut test_buf)) != 0 {
+                    return Some(Err(ParseError::Parse(
+                        format!("Found extra lines after the footer"),
+                        self.current_row_num)));
+                }
+                None
+            },
+            Some(entry) => {
+                if let Entry::Dir(ref dir_path) = entry {
+                    self.current_dir = dir_path.clone();
+                }
+                Some(Ok(entry))
+            },
+        }
+    }
+}
+
+fn read_line<R: BufRead>(reader: &mut R, mut buf: &mut Vec<u8>)
+    -> Result<bool, io::Error>
```

Comment:
No need to return bool here. If you really think file must end in a newline, just turn it into the error immediately and return Error instead of io::Error in this private function.